File: backend/app/api/v1/homework.py
"""
Homework management endpoints
"""

import fastapi
from fastapi import status as http_status

from app.api import dependencies as deps
from app.api import pagination as api_pagination
from app.models import users as user_models
from app.schemas import pagination as pagination_schemas
from app.schemas import homework as homework_schemas
from app.services import homework as homework_service_module
from app.services import testing as testing_service_module

router = fastapi.APIRouter()


@router.post(
    "",
    response_model=homework_schemas.HomeworkResponse,
    status_code=http_status.HTTP_201_CREATED,
)
async def create_homework(
    homework_data: homework_schemas.HomeworkCreate,
    current_user: user_models.User = fastapi.Depends(deps.get_current_teacher),
    homework_service: homework_service_module.HomeworkService = fastapi.Depends(
        deps.get_homework_service
    ),
):
    """
    Create new homework assignment (teachers only)

    - **lesson_id**: ID of the lesson
    - **title**: Homework title
    - **description**: Optional description
    - **max_score**: Maximum score (default 100)
    - **deadline**: Optional deadline date/time
    - **problem_ids**: List of problem IDs to include
    - **problem_points**: Optional list of points for each problem
    """
    return await homework_service.create_homework(homework_data, current_user.id)


@router.get(
    "/lesson/{lesson_id}",
    response_model=pagination_schemas.Page[homework_schemas.HomeworkResponse],
)
async def get_lesson_homework(
    lesson_id: int,
    pagination: api_pagination.Pagination = fastapi.Depends(api_pagination.get_pagination),
    current_user: user_models.User = fastapi.Depends(deps.get_current_user),
    homework_service: homework_service_module.HomeworkService = fastapi.Depends(
        deps.get_homework_service
    ),
):
    """
    Get all homework for a lesson

    - Teachers see all homework (including unpublished)
    - Students see only published homework
    """
    items = await homework_service.get_lesson_homework(
        lesson_id, current_user, pagination.skip, pagination.limit
    )
    total = await homework_service.count_lesson_homework(lesson_id, current_user)
    return pagination_schemas.Page(items=items, total=total, skip=pagination.skip, limit=pagination.limit)


@router.get("/{homework_id}", response_model=homework_schemas.HomeworkDetailResponse)
async def get_homework(
    homework_id: int,
    current_user: user_models.User = fastapi.Depends(deps.get_current_user),
    homework_service: homework_service_module.HomeworkService = fastapi.Depends(
        deps.get_homework_service
    ),
):
    """
    Get homework details by ID

    Returns homework information.
    Students can only see published homework.
    """
    return await homework_service.get_homework(homework_id, current_user)


@router.get(
    "/{homework_id}/problems",
    response_model=list[homework_schemas.ProblemResponse | homework_schemas.ProblemFullResponse],
)
async def get_homework_problems(
    homework_id: int,
    current_user: user_models.User = fastapi.Depends(deps.get_current_user),
    homework_service: homework_service_module.HomeworkService = fastapi.Depends(
        deps.get_homework_service
    ),
):
    """
    Get all problems for homework

    - Teachers see problems with correct answers
    - Students see problems without correct answers
    """
    return await homework_service.get_homework_problems(homework_id, current_user)


@router.post("/submit-answer", response_model=homework_schemas.StatisticsResponse)
async def submit_answer(
    answer_data: homework_schemas.AnswerSubmit,
    current_user: user_models.User = fastapi.Depends(deps.get_current_student),
    testing_service: testing_service_module.TestingService = fastapi.Depends(
        deps.get_testing_service
    ),
):
    """
    Submit answer for a problem (students only).

    Production alias for `/api/v1/testing/submit-answer`.
    """
    return await testing_service.submit_answer(answer_data, current_user.id)


@router.post(
    "/{homework_id}/submit",
    response_model=homework_schemas.StatisticsResponse,
)
async def submit_homework(
    homework_id: int,
    current_user: user_models.User = fastapi.Depends(deps.get_current_student),
    testing_service: testing_service_module.TestingService = fastapi.Depends(
        deps.get_testing_service
    ),
):
    """
    Submit homework for final grading (students only).

    Production alias for `/api/v1/testing/homework/{homework_id}/submit`.
    """
    return await testing_service.submit_homework(homework_id, current_user.id)


@router.get(
    "/{homework_id}/status",
    response_model=homework_schemas.StatisticsResponse,
)
async def get_homework_status(
    homework_id: int,
    current_user: user_models.User = fastapi.Depends(deps.get_current_student),
    testing_service: testing_service_module.TestingService = fastapi.Depends(
        deps.get_testing_service
    ),
):
    """
    Get current status/progress for homework (students only).

    Production alias for `/api/v1/testing/homework/{homework_id}/status`.
    """
    return await testing_service.get_homework_status(homework_id, current_user.id)


@router.patch("/{homework_id}", response_model=homework_schemas.HomeworkResponse)
async def update_homework(
    homework_id: int,
    homework_data: homework_schemas.HomeworkUpdate,
    current_user: user_models.User = fastapi.Depends(deps.get_current_teacher),
    homework_service: homework_service_module.HomeworkService = fastapi.Depends(
        deps.get_homework_service
    ),
):
    """
    Update homework (teachers only, owner only)
    """
    return await homework_service.update_homework(homework_id, homework_data, current_user.id)


@router.delete("/{homework_id}", status_code=http_status.HTTP_204_NO_CONTENT)
async def delete_homework(
    homework_id: int,
    current_user: user_models.User = fastapi.Depends(deps.get_current_teacher),
    homework_service: homework_service_module.HomeworkService = fastapi.Depends(
        deps.get_homework_service
    ),
):
    """
    Delete homework (teachers only, owner only)

    Note: This would need to be implemented in HomeworkService
    """
    # Placeholder - would need implementation
    return None

File: frontend/src/shared/api/homework.ts
import "@/shared/api/openapi";

import {
  HomeworkService,
  TestingService,
  type AnswerSubmit,
  type HomeworkCreate,
  type HomeworkDetailResponse,
  type HomeworkResponse,
  type HomeworkUpdate,
  type Page_HomeworkResponse_,
  type ProblemFullResponse,
  type ProblemResponse,
  type StatisticsResponse,
} from "@/shared/api/generated";
import { clampLimit, clampSkip } from "@/shared/api/pagination";

export const homeworkApi = {
  async listByLesson(lessonId: number, params: { skip?: number; limit?: number } = {}): Promise<Page_HomeworkResponse_> {
    return await HomeworkService.getLessonHomeworkApiV1HomeworkLessonLessonIdGet(
      lessonId,
      clampSkip(params.skip),
      clampLimit(params.limit)
    );
  },

  async get(homeworkId: number): Promise<HomeworkDetailResponse> {
    return await HomeworkService.getHomeworkApiV1HomeworkHomeworkIdGet(homeworkId);
  },

  async create(payload: HomeworkCreate): Promise<HomeworkResponse> {
    return await HomeworkService.createHomeworkApiV1HomeworkPost(payload);
  },

  async update(homeworkId: number, payload: HomeworkUpdate): Promise<HomeworkResponse> {
    return await HomeworkService.updateHomeworkApiV1HomeworkHomeworkIdPatch(homeworkId, payload);
  },

  async remove(homeworkId: number): Promise<void> {
    return await HomeworkService.deleteHomeworkApiV1HomeworkHomeworkIdDelete(homeworkId);
  },

  async getProblems(homeworkId: number): Promise<Array<ProblemResponse | ProblemFullResponse>> {
    return await HomeworkService.getHomeworkProblemsApiV1HomeworkHomeworkIdProblemsGet(homeworkId);
  },

  async submitAnswer(payload: AnswerSubmit): Promise<StatisticsResponse> {
    return await TestingService.submitAnswerApiV1TestingSubmitAnswerPost(payload);
  },

  async submitHomework(homeworkId: number): Promise<StatisticsResponse> {
    return await TestingService.submitHomeworkApiV1TestingHomeworkHomeworkIdSubmitPost(homeworkId);
  },

  async getStatus(homeworkId: number): Promise<StatisticsResponse> {
    return await TestingService.getHomeworkStatusApiV1TestingHomeworkHomeworkIdStatusGet(homeworkId);
  },
} as const;


